src/NuGet.Clients/NuGet.PackageManagement.UI/ViewModels/ReadmePreviewViewModel.cs

```diff
@@ -0,0 +1,90 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.Threading;
+using NuGet.VisualStudio;
+using NuGet.VisualStudio.Internal.Contracts;
+
+namespace NuGet.PackageManagement.UI.ViewModels
+{
+    public sealed class ReadmePreviewViewModel : ViewModelBase
+    {
+        private INuGetPackageFileService _packageFileService;
+
+        public ReadmePreviewViewModel(INuGetPackageFileService packageFileService)
+        {
+            _packageFileService = packageFileService ?? throw new ArgumentNullException(nameof(packageFileService));
+            _errorLoadingReadme = false;
+            _canDetermineReadmeDefined = true;
+            _rawReadme = string.Empty;
+        }
+
+        private bool _errorLoadingReadme;
+
+        public bool ErrorLoadingReadme
+        {
+            get => _errorLoadingReadme;
+            set
+            {
+                SetAndRaisePropertyChanged(ref _errorLoadingReadme, value);
+            }
+        }
+
+        private string _rawReadme;
+
+        public string ReadmeMarkdown
+        {
+            get => _rawReadme;
+            set
+            {
+                SetAndRaisePropertyChanged(ref _rawReadme, value);
+            }
+        }
+
+        private bool _canDetermineReadmeDefined;
+
+        public bool CanDetermineReadmeDefined
+        {
+            get => _canDetermineReadmeDefined;
+            set
+            {
+                SetAndRaisePropertyChanged(ref _canDetermineReadmeDefined, value);
+            }
+        }
+
+        public async Task LoadReadmeAsync(string rawReadmeUrl, CancellationToken cancellationToken)
+        {
+            var newReadmeValue = string.Empty;
+            var isErrorWithReadme = false;
+            bool canDetermineReadmeDefined = false;
+
+            if (!string.IsNullOrWhiteSpace(rawReadmeUrl))
+            {
+
+                await TaskScheduler.Default;
+
+                var readmeStream = await _packageFileService.GetReadmeAsync(new Uri(rawReadmeUrl), cancellationToken);
+                if (readmeStream is not null)
+                {
+                    using StreamReader streamReader = new StreamReader(readmeStream);
+                    var readme = await streamReader.ReadToEndAsync();
+                    if (!string.IsNullOrWhiteSpace(readme))
+                    {
+                        isErrorWithReadme = false;
+                        canDetermineReadmeDefined = true;
+                        newReadmeValue = readme;
+                    }
+                }
+                await NuGetUIThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
```

Comment:
We shouldn't switch to the main thread at the end of this method.

Reply:
I don't think updating properties on a background thread will work with WPF.
Jonatan, what happens if you remove the switch to main thread? I'm guessing either no UI update or an exception?

Reply:
From what I remember I got an exception when I tried to set the view model properties without switching to the main thread. I can remove it and we can see what happens in the next pr, it looks like the unit test passes either way

Reply:
A better way to handle this is to use the [Dispatcher](https://learn.microsoft.com/en-us/dotnet/api/system.windows.threading.dispatcher?view=windowsdesktop-8.0) to check if we're on the main thread or to marshal to the main thread.

Explicitly switching to the main thread is expensive and error-prone because it forces this background operation onto the main thread where it will remain, even when returning to the caller.

See: https://stackoverflow.com/questions/590590/making-%20sure-onpropertychanged-is-called-on-ui-thread-in-mvvm-wpf-app